src/Microsoft.ML.TimeSeries/SequentialAnomalyDetectionTransformBase.cs

```diff
@@ -16,678 +16,676 @@
+                public void CloneState()
                 {
-                    for (i = 1; i < n; ++i)
+                    if (Interlocked.Increment(ref _parent.StateRefCount) > 1)
                     {
-                        diff = RawScoreBuffer[0] - RawScoreBuffer[i];
-                        _sumSquaredDist -= diff * diff;
+                        State = (AnomalyDetectionStateBase)_parent.StateRef.Clone();
                     }
-
-                    diff = RawScoreBuffer[0] - rawScore;
-                    _sumSquaredDist -= diff * diff;
                 }
 
-                return pValue;
+                public ITransformer GetTransformer()
+                {
+                    return _parent;
+                }
             }
-
-            private protected override void SetNaOutput(ref VBuffer<Double> dst)
+            /// <summary>
+            /// The base state class for sequential anomaly detection: this class implements the p-values and martinagle calculations for anomaly detection
+            /// given that the raw anomaly score calculation is specified by the derived classes.
+            /// </summary>
+            internal abstract class AnomalyDetectionStateBase : SequentialTransformerBase<TInput, VBuffer<Double>, TState>.StateBase
             {
-                var outputLength = Parent._outputLength;
-                var editor = VBufferEditor.Create(ref dst, outputLength);
+                // A reference to the parent transform.
+                protected SequentialAnomalyDetectionTransformBase<TInput, TState> Parent;
 
-                for (int i = 0; i < outputLength; ++i)
-                    editor.Values[i] = Double.NaN;
-
-                dst = editor.Commit();
-            }
+                // A windowed buffer to cache the update values to the martingale score in the log scale.
+                private FixedSizeQueue<Double> LogMartingaleUpdateBuffer { get; set; }
 
-            private protected override sealed void TransformCore(ref TInput input, FixedSizeQueue<TInput> windowedBuffer, long iteration, ref VBuffer<Double> dst)
-            {
-                var outputLength = Parent._outputLength;
-                Host.Assert(outputLength >= 2);
+                // A windowed buffer to cache the raw anomaly scores for p-value calculation.
+                private FixedSizeQueue<Single> RawScoreBuffer { get; set; }
 
-                var result = VBufferEditor.Create(ref dst, outputLength);
-                float rawScore = 0;
+                // The current martingale score in the log scale.
+                private Double _logMartingaleValue;
 
-                for (int i = 0; i < outputLength; ++i)
-                    result.Values[i] = Double.NaN;
+                // Sum of the squared Euclidean distances among the raw socres in the buffer.
+                // Used for computing the optimal bandwidth for Kernel Density Estimation of p-values.
+                private Double _sumSquaredDist;
```

Comment:
As you have made so many changes in this file, I guess it's a good time to replace `Double` with `double`.

Reply:
You could also say replace Single with float, right? Good suggestion but clearly not within the scope of this PR where the goal is lockdown the public surface. 

---
In reply to: [254488160](https://github.com/dotnet/machinelearning/pull/2344#discussion_r254488160) [](ancestors = 254488160)

Reply:
I remember that an old decision is that we will not have any task to change Float to float and so on. In the meanwhile, we do need to achieve float and double eventually. Combining the two points, I feel a good timing is a large scale change like this PR.

---
In reply to: [255164804](https://github.com/dotnet/machinelearning/pull/2344#discussion_r255164804) [](ancestors = 255164804,254488160)